### ISSUE TITLE
unicode characters in template expressions

### ISSUE THREAD

**Issue by johan-ohrn** [BorisMoore/jsrender#342]
We have some autogenerated code and for whatever reasons we have javascript objects with property names using non english characters such as åäö.
When we want to print a property from such an object we would type something like this in our template:
```{{:myObj.someÅpropÄerty}}```

This doesn't work because the javascript generated from this template expression is invalid.
```+((v=data.myObj.someÅdata.propÄdata.erty)!=null?v:"");```

I managed to track down the problem to a couple of regexes using \w character class, which doesn't support anything but ascii characters.

I made the following changing in jsrender.js - we use the separate files
Lines 54-57:
```
    uW = "\\w\\u00E4\\u00E5\\u00F6\\u00C4\\u00C5\\u00D6",

    //rPath = /^(!*?)(?:null|true|false|\d[\d.]*|([\w$]+|\.|~([\w$]+)|#(view|([\w$]+))?)([\w$.^]*?)(?:[.[^]([\w$]+)\]?)?)$/g,
    rPath = new RegExp("^(!*?)(?:null|true|false|\\d[\\d.]*|([" + uW + "$]+|\\.|~([" + uW + "$]+)|#(view|([" + uW + "$]+))?)([" + uW + "$.^]*?)(?:[.[^]([" + uW + "$]+)\\]?)?)$", "g"),
    //        not                               object     helper    view  viewProperty pathTokens      leafToken


    //rParams = /(\()(?=\s*\()|(?:([([])\s*)?(?:(\^?)(~?[\w$.^]+)?\s*((\+\+|--)|\+|-|~(?![\w$_])|&&|\|\||===|!==|==|!=|<=|>=|[<>%*:?\/]|(=))\s*|(!*?(@)?[#~]?[\w$.^]+)([([])?)|(,\s*)|(\(?)\\?(?:(')|("))|(?:\s*(([)\]])(?=[.^]|\s*$|[^([])|[)\]])([([]?))|(\s+)/g,
    rParams = new RegExp("(\\()(?=\\s*\\()|(?:([([])\\s*)?(?:(\\^?)(~?[" + uW + "$.^]+)?\\s*((\\+\\+|--)|\\+|-|~(?![" + uW + "$_])|&&|\\|\\||===|!==|==|!=|<=|>=|[<>%*:?\\/]|(=))\\s*|(!*?(@)?[#~]?[" + uW + "$.^]+)([([])?)|(,\\s*)|(\\(?)\\\\?(?:(')|(\"))|(?:\\s*(([)\\]])(?=[.^]|\\s*$|[^([])|[)\\]])([([]?))|(\\s+)", "g"),
```

Seems kinda ugly to hard code the extra non-english characters like this. Perhaps you could make this user configurable or give unicode more generic support?

Even if using non-english characters for property names seem like a bad idea it's still valid javascript and as such I feel the same expression should be valid inside the template.

As a workarround I know we can type
```{{:myObj["someÅpropÄerty"]}}```
but this is ugly.

What's your take on the whole?

**Comment by johan-ohrn** [comment]
I forgot to mention this is on version 0.9.91 but the code is identical in version 1.0.0.

**Comment by BorisMoore** [comment]
Yes, I am in two minds about doing this, given the test matrix, need to document etc. but FWIW the feature would be like this:

Additional code at line 164

	var uW = value && value.addedJsNameChars;
	if (uW) {
		uW = "[\\w" + uW + "$";
		rPath = new RegExp("^(!*?)(?:null|true|false|\\d[\\d.]*|(" + uW + "]+|\\.|~(" + uW + "]+)|#(view|(" + uW + "]+))?)(" + uW + ".^]*?)(?:[.[^](" + uW + "]+)\\]?)?)$", "g");
		rParams = new RegExp("(\\()(?=\\s*\\()|(?:([([])\\s*)?(?:(\\^?)(~?" + uW + ".^]+)?\\s*((\\+\\+|--)|\\+|-|~(?!" + uW + "_])|&&|\\|\\||===|!==|==|!=|<=|>=|[<>%*:?\\/]|(=))\\s*|(!*?(@)?[#~]?" + uW + ".^]+)([([])?)|(,\\s*)|(\\(?)\\\\?(?:(')|(\"))|(?:\\s*(([)\\]])(?=[.^]|\\s*$|[^([])|[)\\]])([([]?))|(\\s+)", "g");
	}
	return value

(Here it is: [jsviews.js.txt](https://github.com/BorisMoore/jsrender/files/2641304/jsviews.js.txt))

Usage:

    $.views.settings.advanced({addedJsNameChars: "\\u00E4\\u00E5\\u00F6\\u00C4\\u00C5\\u00D6"});
    var data = {foo.someÅpropÄerty:444};
    ...

**Comment by BorisMoore** [comment]
Ah, I had missed your comment [above](https://github.com/BorisMoore/jsrender/issues/342#issuecomment-443857654) (the page had not been refreshed when I wrote my response above.)

Yes, I hear you... I'll let it mature a bit before deciding. Meantime, if you can test fully with the update I included above that would be helpful. Let me know if you see any issues/bugs.

**Comment by BorisMoore** [comment]
I have an alternative which I am looking at. I'll probably send you a new patch to test, in the coming days, rather than testing the  version above...

**Comment by johan-ohrn** [comment]
Sounds good. In the meantime I use this quick fix.

**Comment by BorisMoore** [comment]
Here is an update: 
[jsviews.js.txt](https://github.com/BorisMoore/jsrender/files/2651092/jsviews.js.txt)

I managed to figure out a way of allowing all unicode chars, without needing to specifically specify the character codes. The default is not to support unicode chars in names, but you can write

	$.views.settings.advanced({unicodeChars: true}); // support all unicode chars in JavaScript names

or else limit to specific ones:

	$.views.settings.advanced({unicodeChars: "\\u00E4\\u00E5\\u00F6\\u00C4\\u00C5\\u00D6"});

You can't have unicode characters on helpers, or helper properties, tag names or tag properties, etc. Only on properties of data passed in to the link() or render() method.

The generic unicode version, passing true, is done basically by replacing

    /[\w]*/

by 

    /[^\s~/\x00-\x23\x25-\x2D\x3A-\x40[\]\\`{}|]*/

so word characters are any characters other than the specified ones.

It adds 1% to the filesize, which I hesitate about, but at least the usage is easy now. (Assuming it works correctly!) 

So if you are able to test it , it would be helpful, particularly to make sure it does indeed cover all unicode characters as expected (or at least test with a good range of characters...). 

I am still considering whether or not to publish this feature...

**Comment by johan-ohrn** [comment]
I was able to create a test html page. Download [here](https://uploadfiles.io/wnhi7) (sorry about the adds and wait time to download)

I extracted every character from the [regex](https://mathiasbynens.be/demo/javascript-identifier-regex).
The only character that appears to be invalid in fact in chrome is "ⸯ" which I have commented out in the html file.

I tested this with the jsviews version you provided. It throws an error.

I will test some more to see if I can isolate which characters are the culprit.

**Comment by BorisMoore** [comment]
Excellent. That is very  helpful!

I did some testing with that, and all the characters seem to work. There was still an error from JsViews because of `{{:aⸯa}}` still being in the template markup. So I removed it.

Also, of course you need the line:

    $.views.settings.advanced({unicodeChars: true});

before calling `$.templates(...)`

I also had  some issues when using a file that did not have the BOM (with UTF-8 encoding).

But with those changes, the template rendered completely...

**Comment by BorisMoore** [comment]
Here is an updated jsviews.js and jsrender.js and a copy of my version of your test page:

[jsviews.js.txt](https://github.com/BorisMoore/jsrender/files/2656016/jsviews.js.txt)
[jsrender.js.txt](https://github.com/BorisMoore/jsrender/files/2656017/jsrender.js.txt)
[testUnicode.html.txt](https://github.com/BorisMoore/jsrender/files/2656026/testUnicode.html.txt)

The test page works for me in Chrome and in Firefox, but not in IE, Edge, Safari... It seems that some of the unicode characters trigger javascript errors in those browsers. With a reduced set of unicode characters it works in all those browsers.

**Comment by johan-ohrn** [comment]
I was at the end of my shift and had been messing back and forth with this test page so I missed the obvious $.views.settings.advanced(...) call :)

Good that you made it work!

So we confirmed that the regex does indeed cover all unicode characters.
The fact that it covers more than some browsers doesn't really matter I think.

Also since the regex does cover all unicode characters I don't see the need for a user to explicitly allow specific characters. Unicode on or off is sufficient. This should also result in a lower footprint for the feature.

If the option is to allow unicode or not then I see two possible implementations, well three.
1) Making unicode opt-in via ```$.views.settings.advanced({unicodeChars: true});```
This alternative should ensure that no existing templates are broken by an unexpected bug in the new regex. Also adds more bytes to the final js file.
2) Somehow allow setting the full rPath and rParams regexes from outside jsviews.js. This would allow you to make unicode support opt-in from a jsviews-unicode.js file moving the overhead completely from jsviews.js itself.
3) Skip the option alltogether and just change the regex thus making unicode supported out of the box.

What's your take on these ideas?

**Comment by BorisMoore** [comment]
Yes, I agree with your analysis. #2 is probably my preference. #3 makes me nervous about possible bugs, especially doing it just after releasing v1.0.0 which is hopefully stable. So #2 is a more conservative approach, given that this feature is unlikely to be used/needed by many users...

I'll post a version of #2 soon.

**Comment by BorisMoore** [comment]
Here is an update where you load the jsrender-unicode.js file after jsrender.js/jsviews.js, before calling `$.templates(...)` to compile any templates that have unicode characters in them.

Let me know if it works as expected. Thanks!

[jsrender-unicode.js.txt](https://github.com/BorisMoore/jsrender/files/2660922/jsrender-unicode.js.txt)
[jsrender.js.txt](https://github.com/BorisMoore/jsrender/files/2660923/jsrender.js.txt)
[jsviews.js.txt](https://github.com/BorisMoore/jsrender/files/2660924/jsviews.js.txt)

**Comment by johan-ohrn** [comment]
I gave it a try but run into some issues with jsrender-unicode.js

The regexes in jsrender-unicode.js is not the same as those we used [here](https://github.com/BorisMoore/jsrender/issues/342#issuecomment-445134319)
Is this intentional?
Modifying jsrender-unicode.js to use those regexes seem to work (comments left out for abbrevity):
```
$.views.sub.rPath =
    /^(!*?)(?:null|true|false|\d[\d.]*|([^.^\s~/\x00-\x23\x25-\x2D\x3A-\x40[\]\\`{}|]+|\.|~([\w$]+)|#(view|([\w$]+))?)([^\s~/\x00-\x23\x25-\x2D\x3A-\x40[\]\\`{}|]*?)(?:[.[^]([^.^\s~/\x00-\x23\x25-\x2D\x3A-\x40[\]\\`{}|]+)\]?)?)$/g;
$.views.sub.rPrm =
  /(\()(?=\s*\()|(?:([([])\s*)?(?:(\^?)(~?[\w$.^]+)?\s*((\+\+|--)|\+|-|~(?![\w$])|&&|\|\||===|!==|==|!=|<=|>=|[<>%*:?/]|(=))\s*|(!*?(@)?[#~]?[^\s~/\x00-\x23\x25-\x2D\x3A-\x40[\]\\`{}|]+)([([])?)|(,\s*)|(\(?)\\?(?:(')|(\"))|(?:\s*(([)\]])(?=[.^]|\s*$|[^([])|[)\]])([([]?))|(\s+)/g;
```

I also found some left over code and double comments (line numbers from jsrender.js):
line 114:
```// subscription, e.g. JsViews integration```
line 161:
```
				var unicodeChars = value && value.unicodeChars;
				if (unicodeChars) {
					setPathRegEx(unicodeChars);
				}
```

You might want to double check any changes related to this issue.

**Comment by BorisMoore** [comment]
Ah - sorry, I must have accidentally uploaded an intermediate version. Thanks for pointing out the stale code/comments, too!

So hopefully this is correct now. The regEx expressions are close to yours, but not quite identical. Let me know if you have any concerns, or see any issues.

[jsrender-unicode.js.txt](https://github.com/BorisMoore/jsrender/files/2664498/jsrender-unicode.js.txt)
[testUnicode.html.txt](https://github.com/BorisMoore/jsrender/files/2664499/testUnicode.html.txt)
[jsviews.js.txt](https://github.com/BorisMoore/jsrender/files/2664502/jsviews.js.txt)
[jsrender.js.txt](https://github.com/BorisMoore/jsrender/files/2664503/jsrender.js.txt)

**Comment by johan-ohrn** [comment]
I finally got to implement the fix in our code. Works perfectly so far!
I'll test it some more and see if anything turns up.

**Comment by BorisMoore** [comment]
I'll include this in the next update  (v1.0.1)

**Comment by johan-ohrn** [comment]
Glad to hear.
No problems so far.

**Comment by BorisMoore** [comment]
Fixed in commit v1.0.1
